File: Managed/NextTurn.UE.Runtime/Core/SlowTask.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public class SlowTask : IDisposable
    {
        private readonly IntPtr pointer;

        private bool disposed = false;

        public SlowTask(float totalAmount, Text? defaultMessage = null)
        {
            defaultMessage ??= Text.Empty;
            this.pointer = NativeMethods.New(totalAmount, defaultMessage.text);
        }

        ~SlowTask() => this.Dispose(false);

        public void Dispose()
        {
            this.Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!this.disposed)
            {
                NativeMethods.Delete(this.pointer);

                this.disposed = true;
            }
        }

        private static class NativeMethods
        {
            [ReadOffset]
            public static extern float GetCompletedAmount(IntPtr task);

            [ReadOffset]
            public static extern NativeText GetDefaultMessage(IntPtr task);

            [ReadOffset]
            public static extern NativeText GetFrameMessage(IntPtr task);

            [ReadOffset]
            public static extern float GetTotalAmount(IntPtr task);

            [Calli]
            public static extern IntPtr New(float totalAmount, in NativeText defaultMessage);

            [Calli]
            public static extern void Delete(IntPtr task);
        }
    }
}

File: Source/NextTurnRuntime/Private/Core/SlowTask.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/Accessor.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"
#include "Runtime/Core/Public/Misc/SlowTask.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME SlowTask

TYPE_EXPORT_START

FSlowTask* EXPORT_CALL_CONV New(float TotalAmount, const FText& DefaultMessage)
{
	FSlowTask* Result = new FSlowTask(TotalAmount, DefaultMessage);
	Result->Initialize();
	return Result;
}

void EXPORT_CALL_CONV Delete(FSlowTask* Task)
{
	Task->Destroy();
	delete Task;
}

const FMemberSymbol Members[] =
{
	EXPORT_OFFSET_PUBLIC(FSlowTask, CompletedAmount, CompletedWork),
	EXPORT_OFFSET_PUBLIC(FSlowTask, DefaultMessage, DefaultMessage),
	EXPORT_OFFSET_PUBLIC(FSlowTask, FrameMessage, FrameMessage),
	EXPORT_OFFSET_PUBLIC(FSlowTask, TotalAmount, TotalAmountOfWork),

	EXPORT_METHOD(New),
	EXPORT_METHOD(Delete),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
